contracts/OGNRewardsSource.sol

```diff
@@ -0,0 +1,156 @@
+
+    address public strategistAddr;
+
+    address public rewardsTarget;
+
+    struct RewardConfig {
+        // Inspired by (Copied from) `Dripper.Drip` struct.
+        uint64 lastCollect; // Overflows 262 billion years after the sun dies
+        uint192 rewardsPerSecond;
+    }
+
+    RewardConfig public rewardConfig;
+
+    /**
+     * @dev Verifies that the caller is either Governor or Strategist.
+     */
+    modifier onlyGovernorOrStrategist() {
+        if (msg.sender != strategistAddr && !isGovernor()) {
+            revert UnauthorizedCaller();
+        }
+
+        _;
+    }
+
+    constructor(address _ogn) {
+        ogn = _ogn;
+    }
+
+    /// @dev Initialize the proxy implementation
+    /// @param _strategistAddr Address of the Strategist
+    /// @param _rewardsTarget Address that receives rewards
+    /// @param _rewardsPerSecond Rate of reward emission
+    function initialize(address _strategistAddr, address _rewardsTarget, uint256 _rewardsPerSecond)
+        external
+        initializer
+    {
+        _setStrategistAddr(_strategistAddr);
+        _setRewardsTarget(_rewardsTarget);
+
+        // Rewards start from the moment the contract is initialized
+        rewardConfig.lastCollect = uint64(block.timestamp);
+
+        _setRewardsPerSecond(_rewardsPerSecond);
+    }
+
+    /// @dev Collect pending rewards
+    /// @return rewardAmount Amount of reward collected
+    function collectRewards() external returns (uint256) {
+        return _collectRewards();
+    }
+
+    /// @dev Collect pending rewards
+    /// @return rewardAmount Amount of reward collected
+    function _collectRewards() internal returns (uint256 rewardAmount) {
+        address _target = rewardsTarget;
+        if (_target == address(0)) {
+            revert RewardsTargetNotSet();
+        }
+
+        // Compute pending rewards
+        RewardConfig storage _config = rewardConfig;
+        rewardAmount = _previewRewards(_config);
+
+        // Update timestamp
+        _config.lastCollect = uint64(block.timestamp);
+
+        if (rewardAmount > 0) {
+            // Should not revert if there's no reward to transfer.
+
+            emit RewardCollected(rewardAmount);
+
+            // Intentionally skipping balance check to save some gas
+            // since `transfer` anyway would fail in case of low balance
+            IERC20(ogn).transfer(_target, rewardAmount);
+        }
+    }
+
+    /// @dev Compute pending rewards since last collect
+    /// @return rewardAmount Amount of reward that'll be distributed if collected now
+    function previewRewards() external view returns (uint256) {
+        return _previewRewards(rewardConfig);
+    }
+
+    /// @dev Compute pending rewards since last collect
+    /// @param _rewardConfig RewardConfig
+    /// @return rewardAmount Amount of reward that'll be distributed if collected now
+    function _previewRewards(RewardConfig memory _rewardConfig) internal view returns (uint256) {
```

Comment:
Is there any difference in gas usage from having an internal / and external _preview rewards, vs one public method?

Reply:
Will run some tests and see if it makes any change

Reply:
Having a public method was slightly cheaper in terms of gas usage. Have changed the code